{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7365e4a8-77ea-41d3-981e-d909f111e64f",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(Seurat)\n",
    "library(SeuratDisk)\n",
    "library(ggplot2)\n",
    "library(reticulate)\n",
    "library(SingleCellExperiment)\n",
    "library(edgeR)\n",
    "library(scran)\n",
    "library(ggplot2)\n",
    "library(ggrepel)\n",
    "library(clusterProfiler)\n",
    "library(enrichplot)\n",
    "organism = \"org.Hs.eg.db\"\n",
    "BiocManager::install(organism, character.only = TRUE)\n",
    "library(organism, character.only = TRUE)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8a7b7d7-5284-4a91-ba31-e49191ceccda",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Convert H5AD to Seurat\n",
    "Convert(\"./combined_CRX_40um_plaqueDistances_forDE.h5ad\", dest = \"h5seurat\", overwrite = TRUE)\n",
    "seurat_obj <- LoadH5Seurat(\"./combined_CRX_40um_plaqueDistances_forDE.h5seurat\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23469cd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create dictionary mapping to name smaples\n",
    "mapping <- c(\"B1S3\" = \"LEC_1\", \"B1S4\" = \"LALA-PG_1\", \"B2S3\" = \"LEC_2\", \"B2S4\" = \"LALA-PG_2\")\n",
    "\n",
    "seurat_obj@meta.data$SampleID <- mapping[seurat_obj@meta.data$sample]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63000dca-9384-4dc0-9c07-611d65c8fa53",
   "metadata": {},
   "outputs": [],
   "source": [
    "x <- 90  # Set TD count threshold\n",
    "\n",
    "seurat_obj_sub <- subset(seurat_obj, features = rownames(seurat_obj)[Matrix::rowSums(seurat_obj@assays$RNA@counts) >= x])\n",
    "# set distance threshold \n",
    "seurat_obj_sub = subset(seurat_obj_sub, subset = distance_to_plaques < 200)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98b41249-bbe5-41c4-86f9-d491abca41a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#define funciton for plotting MAs of results of edgeR DEs\n",
    "\n",
    "MAplot <- function(resTable, overlaps, force = 2) {\n",
    "  \n",
    "  resTable2 <- resTable[order(resTable$logFC), ]\n",
    "  resTable2$gene <- rownames(resTable2)\n",
    "  resTable2$threshold <- as.factor(resTable2$FDR < 0.05)\n",
    "  \n",
    "  # Aesthetics\n",
    "  plot_aesthetics <- list(\n",
    "    xlab(\"\\nLog2 Counts Per Million (logCPM)\"),\n",
    "    ylab(\"Log2 Fold Change (logFC)\\n\"),\n",
    "    theme_minimal(base_size = 15),\n",
    "    theme(\n",
    "      axis.title.x = element_text(face = \"bold\", size = 18, color = \"black\"),\n",
    "      axis.text.x = element_text(size = 14, color = \"black\"),\n",
    "      axis.title.y = element_text(face = \"bold\", size = 18, color = \"black\"),\n",
    "      axis.text.y = element_text(size = 14, color = \"black\"),\n",
    "      legend.title = element_text(face = \"bold\", size = 15),\n",
    "      legend.text = element_text(size = 13),\n",
    "      plot.title = element_text(face = \"bold\", size = 20, hjust = 0.5, color = \"black\"),\n",
    "      panel.grid.major = element_blank(),\n",
    "      panel.grid.minor = element_blank(),\n",
    "      plot.background = element_rect(fill = \"white\", color = NA),\n",
    "      legend.background = element_rect(fill = \"white\", color = NA)\n",
    "    ),\n",
    "    scale_colour_manual(name = \"p.adjusted < 0.05\", values = c(\"steelblue\", \"red2\")),\n",
    "    scale_size_manual(name = \"p.adjusted < 0.05\", values = c(1.2, 2))\n",
    "  )\n",
    "  \n",
    "  #  MA plot \n",
    "  ma_plot <- ggplot(data = resTable2, aes(x = logCPM, y = logFC, colour = threshold, size = threshold)) +\n",
    "    geom_point(aes(alpha = ifelse(threshold == \"TRUE\", 0.9, 0.6)), show.legend = FALSE) +\n",
    "    geom_hline(aes(yintercept = 0), colour = \"blue\", size = 0.8, linetype = \"dashed\") +\n",
    "    ylim(c(min(resTable2$logFC), max(resTable2$logFC))) + \n",
    "    plot_aesthetics\n",
    "  \n",
    "  # Add gene labels for X non-significant genes\n",
    "  ma_plot <- ma_plot + \n",
    "    geom_label_repel(\n",
    "      data = subset(head(resTable2[order(-abs(resTable2$logFC)), ], 30), FDR > 0.05 & logCPM > 14.8),\n",
    "      aes(logCPM, logFC, label = gene),\n",
    "      size = 4,\n",
    "      color = \"steelblue\",  \n",
    "      force = force,\n",
    "      box.padding = 0.5,\n",
    "      max.overlaps = overlaps,\n",
    "      segment.color = \"gray50\",\n",
    "      segment.size = 0.5\n",
    "    )\n",
    "  \n",
    "  # Add gene labels for significant genes \n",
    "  ma_plot <- ma_plot + \n",
    "    geom_label_repel(\n",
    "      data = subset(resTable2, FDR < 0.05),\n",
    "      aes(logCPM, logFC, label = gene),\n",
    "      size = 5,\n",
    "      color = \"red2\",  \n",
    "      force = force, \n",
    "      max.overlaps = overlaps,\n",
    "      box.padding = 0.6,\n",
    "      segment.color = \"gray50\", \n",
    "      segment.size = 0.5\n",
    "    )\n",
    "  \n",
    "  ma_plot <- ma_plot + ggtitle(\"MA Plot of Differential Expression\")\n",
    "  \n",
    "  return(ma_plot)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d476d713-e1e7-4e6a-a4cd-ef0a16c20cfc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "seurat_obj_lec = subset(seurat_obj_sub, subset = sample %in% c('B2S3', 'B1S3'))\n",
    "\n",
    "counts <- as.matrix(seurat_obj_lec@assays$RNA@counts)\n",
    "metadata <- seurat_obj_lec@meta.data\n",
    "\n",
    "sce <- SingleCellExperiment(assays = counts, \n",
    "                                    colData = metadata)\n",
    "\n",
    "dge <- convertTo(sce, type=\"edgeR\", assay.type = 1)\n",
    "dge <- calcNormFactors(dge)\n",
    "\n",
    "meta_dge <- dge$samples\n",
    "meta_dge <- meta_dge[,c(\"lib.size\",\"norm.factors\")]\n",
    "\n",
    "#specify design for  glm\n",
    "design <- model.matrix(~metadata$distance_to_plaques)\n",
    "\n",
    "dge <- estimateDisp(dge, design = design)\n",
    "fit <- glmQLFit(dge, design = design)\n",
    "qlf <- glmQLFTest(fit)\n",
    "tt <- topTags(qlf, n = Inf)\n",
    "\n",
    "res_lec = tt$table\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28b955f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "restable_lec = res_lec\n",
    "\n",
    "options(repr.plot.width=20, repr.plot.height=14)\n",
    "restable_lec$logFC = restable_lec$logFC*-1 #invert LFCs so positive indicates closer to plaques\n",
    "restable_lec$logFC = restable_lec$logFC*100\n",
    "\n",
    "MAplot(restable_lec, 14, force = 0.2) + ggtitle(\"CRX LEC: 40um hexbins with respect to distance to AB\") + \n",
    "    labs(subtitle = \"LFC for 100um closer to pathology\")  +\n",
    "    xlim(min(restable$logCPM), 15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bba3905b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d05e673-cec8-498c-9eec-a047b459a78a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "seurat_obj_lala = subset(seurat_obj_sub, subset = sample %in% c('B1S4', 'B2S4'))\n",
    "\n",
    "\n",
    "counts <- as.matrix(seurat_obj_lala@assays$RNA@counts)\n",
    "metadata <- seurat_obj_lala@meta.data\n",
    "\n",
    "\n",
    "sce <- SingleCellExperiment(assays = counts, \n",
    "                                    colData = metadata)\n",
    "\n",
    "dge <- convertTo(sce, type=\"edgeR\", assay.type = 1)\n",
    "dge <- calcNormFactors(dge)\n",
    "\n",
    "meta_dge <- dge$samples\n",
    "meta_dge <- meta_dge[,c(\"lib.size\",\"norm.factors\")]\n",
    "\n",
    "design <- model.matrix(~metadata$distance_to_plaques)\n",
    "dge <- estimateDisp(dge, design = design)\n",
    "fit <- glmQLFit(dge, design = design)\n",
    "qlf <- glmQLFTest(fit)\n",
    "tt <- topTags(qlf, n = Inf)\n",
    "res_lala = tt$table\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "480a2cd1-0e3c-4eb7-8a58-7351e4e1e781",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ed86016-58fd-4452-9ccc-8ab66f741ea8",
   "metadata": {},
   "outputs": [],
   "source": [
    "restable_lala = res_lala\n",
    "\n",
    "options(repr.plot.width=20, repr.plot.height=14)\n",
    "restable_lala$logFC = restable_lala$logFC*-1\n",
    "restable_lala$logFC = restable_lala$logFC*100\n",
    "\n",
    "MAplot(restable_lala, 15, force = 0.2) + ggtitle(\"CRX LALA: 40um hexbins with respect to distance to AB\") + \n",
    "    labs(subtitle = \"LFC for 100um closer to pathology\")  +\n",
    "    xlim(min(restable$logCPM), 15.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bca7465",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotQuadrant <- function(results_1, results_2, title) {\n",
    "  \n",
    "  # Merge the two dataframes by  'Gene' \n",
    "  merged_df <- merge(results_1, results_2, by = \"X\", suffixes = c(\"_1\", \"_2\"))\n",
    "  \n",
    "  # indicate Significance\n",
    "  merged_df$Significance <- with(merged_df, \n",
    "    ifelse(FDR_1 < 0.05 & FDR_2 < 0.05, \"Both significant\",\n",
    "           ifelse(FDR_1 < 0.05, \"Significant in LALA-PG only\",\n",
    "           ifelse(FDR_2 < 0.05, \"Significant in Lec only\", \"Neither significant\"))))\n",
    "\n",
    "  merged_df$label_grey <- with(merged_df, Significance == \"Neither significant\")\n",
    "\n",
    "  # Aesthetic\n",
    "  base_plot <- ggplot(merged_df, aes(x = logFC_1, y = logFC_2)) +\n",
    "    geom_vline(xintercept = 0, linetype = \"dashed\", color = \"black\") +\n",
    "    geom_hline(yintercept = 0, linetype = \"dashed\", color = \"black\") +\n",
    "    geom_abline(slope = 1, intercept = 0, linetype = \"dashed\", color = \"red\")\n",
    "\n",
    "  # Add points\n",
    "  plot_with_points <- base_plot + \n",
    "    geom_point(data = merged_df[merged_df$Significance != \"Neither significant\",], \n",
    "               aes(color = Significance), size = 3, alpha = 0.8) +\n",
    "    geom_point(data = merged_df[merged_df$Significance == \"Neither significant\",], \n",
    "               aes(color = Significance), size = 0.5, alpha = 0.15)\n",
    "\n",
    "  # Add labels \n",
    "  plot_with_labels <- plot_with_points +\n",
    "    geom_label_repel(data = merged_df[merged_df$Significance == \"Significant in LALA-PG only\",], \n",
    "                     aes(label = X), size = 8, color = \"purple2\", \n",
    "                     label.size = 0.5, max.overlaps = 7) +\n",
    "    geom_label_repel(data = merged_df[merged_df$Significance == \"Significant in Lec only\",], \n",
    "                     aes(label = X), size = 9, color = \"red2\", \n",
    "                     label.size = 0.5, max.overlaps = 18) +\n",
    "    geom_label_repel(data = merged_df[merged_df$Significance == \"Both significant\",], \n",
    "                     aes(label = X), size = 9, color = 'darkgreen', \n",
    "                     label.size = 0.5, max.overlaps = 12)\n",
    "\n",
    "  final_plot <- plot_with_labels + \n",
    "    scale_color_manual(values = c(\"Both significant\" = \"darkgreen\", \n",
    "                                  \"Significant in LALA-PG only\" = \"purple2\", \n",
    "                                  \"Significant in Lec only\" = \"red2\", \n",
    "                                  \"Neither significant\" = \"darkgray\"),\n",
    "                       limits = c(\"Both significant\", \"Significant in Lec only\", \"Significant in LALA-PG only\", \"Neither significant\")) +\n",
    "    theme_minimal(base_size = 26) +\n",
    "    theme(plot.margin = unit(c(1, 1, 1, 1), \"lines\")) +\n",
    "    ggtitle(label = title)\n",
    "  \n",
    "  # Return  final plot\n",
    "  return(final_plot)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdc03bd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=22, repr.plot.height=15)\n",
    "\n",
    "restable_lala$X = gsub( \"GRCh38-\", \"\", row.names(restable_lala))\n",
    "restable_lec$X = gsub( \"GRCh38-\", \"\", row.names(restable_lec))\n",
    "\n",
    "\n",
    "plotQuadrant(restable_lala, restable_lec, 'DE of TDs with respect to distance to pathology (cortex)') + \n",
    "    xlab(\"\\n Lec LALA-PG (LFC for 100um closer to pathology)\") + ylab(\"Lec (LFC for 100um closer to pathology)\\n \") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d910601f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e9281a2a",
   "metadata": {},
   "source": [
    "### KEGG GSEA on DEs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aecf6b6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# Function to preprocess and perform GSEA KEGG\n",
    "perform_GSEA_KEGG <- function(df, title, colour_up, colour_down, colour_ns) {\n",
    "\n",
    "    df$Gene <- sub(\"GRCh38-\", \"\", rownames(df))\n",
    "    gene_list <- sort(na.omit(df$logFC), decreasing = TRUE)\n",
    "    \n",
    "    # Convert gene symbols to ENTREZ IDs\n",
    "    ids <- bitr(names(gene_list), fromType = \"SYMBOL\", toType = \"ENTREZID\", OrgDb = org.Hs.eg.db)\n",
    "    dedup_ids <- ids[!duplicated(ids$SYMBOL),]\n",
    "    df2 <- df[df$Gene %in% dedup_ids$SYMBOL,]\n",
    "    df2$Y <- dedup_ids$ENTREZID\n",
    "    \n",
    "    # Prepare KEGG gene list\n",
    "    kegg_gene_list <- sort(na.omit(df2$logFC), decreasing = TRUE)\n",
    "    names(kegg_gene_list) <- df2$Y\n",
    "    \n",
    "    # Perform GSEA KEGG\n",
    "    kk_result <- gseKEGG(geneList = kegg_gene_list, organism = \"hsa\", nPerm = 1000,\n",
    "                         minGSSize = 70, maxGSSize = 800, pvalueCutoff = 0.05,\n",
    "                         pAdjustMethod = \"BH\", keyType = \"ncbi-geneid\")\n",
    "    \n",
    "    res <- kk_result@result\n",
    "    res$Treatment <- ifelse(res$NES > 0 & res$qvalue < 0.05, \"Up\",\n",
    "                            ifelse(res$NES < 0 & res$qvalue < 0.05, \"Down\", \"Not Sig\"))\n",
    "    \n",
    "    res$Significance <- ifelse(res$qvalue < 0.001, \"***\",\n",
    "                                ifelse(res$qvalue < 0.01, \"**\", ifelse(res$qvalue < 0.05, \"*\", \"\")))\n",
    "    \n",
    "    # Create plot\n",
    "    ggplot(head(res, 30), aes(x = reorder(Description, -NES), y = NES, fill = Treatment, color = Treatment)) +\n",
    "        geom_bar(stat = \"identity\", width = 0.75, alpha = 0.5, size = 1.5) +\n",
    "        geom_text(aes(label = Significance, y = NES - 0.2 * sign(NES)), size = 8, fontface = \"bold\") +\n",
    "        scale_fill_manual(values = c(\"Up\" = colour_up, \"Down\" = colour_down, \"Not Sig\" = colour_ns)) +\n",
    "        scale_color_manual(values = c(\"Up\" = colour_up, \"Down\" = colour_down, \"Not Sig\" = colour_ns)) +\n",
    "        scale_x_discrete(expand = expansion(add = .5)) +\n",
    "        scale_y_continuous(breaks = seq(floor(min(res$NES)), ceiling(max(res$NES)),\n",
    "                                       ceiling((ceiling(max(res$NES)) - floor(min(res$NES))) / 6))) +\n",
    "        coord_flip() + \n",
    "        theme_bw(base_size = 29) +\n",
    "        theme(plot.title = element_text(size = 23), legend.position = \"none\",\n",
    "              plot.margin = margin(10, 10, 10, 35)) +\n",
    "        ggtitle(title) +\n",
    "        xlab(\"\")\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43336eee",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Perform GSEA KEGG for lec DE\n",
    "perform_GSEA_KEGG(\n",
    "  restable_lec, \n",
    "  \"LEC: GSEA KEGG wrt distance to pathology (cortex)\", \n",
    "  colour_up = \"red\", colour_down = \"darkgrey\", colour_ns = \"#A9A9A9\"\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01627308",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform GSEA KEGG for lala de\n",
    "perform_GSEA_KEGG(\n",
    "  restable_lala, \n",
    "  \"LALA: GSEA KEGG wrt distance to pathology (cortex)\", \n",
    "  colour_up = \"purple2\", colour_down = \"darkgrey\", colour_ns = \"#A9A9A9\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fe5cfc4",
   "metadata": {},
   "source": [
    "## Enrichment of scRNA-seq hdWGCNA modules "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c368aca0",
   "metadata": {},
   "outputs": [],
   "source": [
    "modules = read.csv('../../../scRNA/900.wgcna/modules_hdWGCNA_lecMicro.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e70a366",
   "metadata": {},
   "outputs": [],
   "source": [
    "#prepare modules for GSEA format \n",
    "\n",
    "# Create an empty list to store individual module data frames\n",
    "module_list <- list()\n",
    "\n",
    "# Iterate over each unique module\n",
    "for(module_name in unique(modules$module)) {\n",
    "  # Extract genes for  current module\n",
    "  module_genes <- modules[modules$module == module_name, c(\"gene_name\")]\n",
    "  \n",
    "  # Create a data frame for current module\n",
    "  module_df <- data.frame(geneID = module_genes, set = module_name)\n",
    "  \n",
    "  # Append the module data frame to the list\n",
    "  module_list[[module_name]] <- module_df\n",
    "}\n",
    "\n",
    "sets <- do.call(rbind, module_list)\n",
    "row.names(sets) = NULL\n",
    "\n",
    "colnames(sets) <- c(\"geneID\", \"set\")\n",
    "sets2 <- data.frame(cbind(sets$set, sets$geneID))\n",
    "colnames(sets2) <- c(\"set\", \"geneID\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "370ac93d",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=8)\n",
    "\n",
    "df = restable_lec\n",
    "df$Gene <- sub(\"GRCh38-\", \"\", rownames(df))  # Remove GRCh38- prefix\n",
    "original_gene_list <-  df$logFC \n",
    "names(original_gene_list) <- df$Gene\n",
    "gene_list<-na.omit(original_gene_list)\n",
    "gene_list = sort(gene_list, decreasing = TRUE)\n",
    "\n",
    "\n",
    "gsea_mod_lec <- GSEA(gene_list, TERM2GENE = sets2, eps = 0,minGSSize = 5, maxGSSize = 800,\n",
    "             pAdjustMethod = \"BH\",  pvalueCutoff = 0.05)\n",
    "\n",
    "gsea_mod_lec@result = gsea_mod_lec@result[order(gsea_mod_lec@result$NES),]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37c9aa6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=12, repr.plot.height=10)\n",
    "\n",
    "# Assuming res$Description contains color names or hex codes\n",
    "res = gsea_mod_lec@result\n",
    "\n",
    "res$Treatment = ifelse(res$NES > 0 & res$p.adjust < 0.05, \"Up near plaques\", \n",
    "                   ifelse(res$NES < 0 & res$p.adjust < 0.05, \"Down near plaques\", \"C\"))\n",
    "\n",
    "\n",
    "res$OutlineColor <- ifelse(res$NES > 0, \"Up near plaques\", \"Down near plaques\")\n",
    "res$OutlineColor <- factor(res$OutlineColor, levels = c(\"Up near plaques\", \"Down near plaques\"))\n",
    "\n",
    "\n",
    "# Define significance levels\n",
    "res$Significance <- ifelse(res$qvalue < 0.001, \"***\",\n",
    "                     ifelse(res$qvalue < 0.01, \"**\",\n",
    "                     ifelse(res$qvalue < 0.05, \"*\", \"\")))\n",
    "\n",
    "\n",
    "ggplot(res, aes(x = reorder(ID, -NES), y = NES, fill = Description, color = OutlineColor)) +\n",
    "      geom_bar(stat = \"identity\", width = 0.8, size = 1.4, alpha = 0.75) + \n",
    "      geom_text(aes(label = Significance, y = NES - 0.3 * sign(NES)), \n",
    "                size = 6, fontface = \"bold\") +\n",
    "    geom_hline(yintercept = 0,size = 1)+\n",
    "      scale_fill_identity() +  # Use colors from Description column\n",
    "      scale_color_manual(values = c(\"Up near plaques\" = \"red2\", \"Down\" = \"#7a7a7a\")) +  \n",
    "      \n",
    "      scale_x_discrete(expand = expansion(add = .5)) +\n",
    "      scale_y_continuous(\n",
    "        breaks = seq(\n",
    "          floor(min(res$NES)), ceiling(max(res$NES)),\n",
    "          ceiling((ceiling(max(res$NES)) - floor(min(res$NES))) / 6)\n",
    "        )       ) +\n",
    "      coord_flip() + \n",
    "      theme_bw(base_size = 22) +   \n",
    "        theme(\n",
    "        legend.position = \"right\",\n",
    "        legend.key.size = unit(1.5, \"lines\"),\n",
    "        legend.text = element_text(size = 23),\n",
    "        axis.text = element_text(size = 25),\n",
    "        axis.title = element_text(size = 25), \n",
    "        plot.title = element_text(size = 27, face = 'bold', hjust = 0.6, \n",
    "                                 margin = margin(b = 10)), \n",
    "        plot.margin = margin(10,120,10, 10), \n",
    "        legend.title = element_text(size = 24)\n",
    "\n",
    "      ) +  \n",
    "      xlab(\"WGCNA Modules\\n\")+\n",
    "    labs(title = \"GSEA: Lec TDs with respect to\\n distance to pathology\",  y = \"NES\", color = \"Direction\") +  # Change the legend title to \"NES Direction\"\n",
    "      labs(color = NULL, fill = NULL) +  \n",
    "    labs(fill = NULL) + \n",
    "      guides(fill = \"none\", color = 'none') \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecc40c9b-dfbb-4af6-8f58-3a834fe7133d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r4_012024",
   "language": "R",
   "name": "r4_012024"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
