{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73304b66",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(Seurat)\n",
    "library(tidyverse)\n",
    "library(cowplot)\n",
    "library(patchwork)\n",
    "library(ggplot2)\n",
    "library(sWGCNA)\n",
    "library(hdWGCNA)\n",
    "library(dplyr)\n",
    "library(doParallel)\n",
    "library(edgeR)\n",
    "library(ggpubr) \n",
    "library(lsr) \n",
    "library(viridis)\n",
    "library(Hmisc)\n",
    "library(pheatmap)\n",
    "library(grid)\n",
    "library(RColorBrewer)\n",
    "library(gridExtra)\n",
    "library(clusterProfiler)\n",
    "library(GO.db)\n",
    "library(\"org.Hs.eg.db\")\n",
    "library(ggrepel)\n",
    "library(ggraph)\n",
    "library(tidygraph)\n",
    "library(stringr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68a2883a",
   "metadata": {},
   "source": [
    "## Prepare hdWGCNA metacell object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6b9e4ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "seurat_obj = readRDS('./scrna_lecMicro_cleaned_integrated_sctransform.RDS')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "860001d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "DefaultAssay(seurat_obj) = 'RNA'\n",
    "\n",
    "seurat_obj <- SetupForWGCNA(\n",
    "  seurat_obj,\n",
    "  gene_select = \"fraction\", \n",
    "  fraction = 0.05, # fraction of cells that a gene needs to be expressed\n",
    "  wgcna_name = \"lecMicro\" \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b76ab8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# construct metacells  in each group\n",
    "seurat_obj <- MetacellsByGroups(target_metacells = 200,\n",
    "  seurat_obj = seurat_obj,\n",
    "  group.by = c(\"hash.ID\"), # to group by\n",
    "  reduction = 'pca', # dimensionality reduction to perform KNN on\n",
    "  k = 25,\n",
    "  max_shared = 10, \n",
    "  ident.group = 'hash.ID' #  Idents of the metacell seurat object\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdc62813",
   "metadata": {},
   "outputs": [],
   "source": [
    "seurat_obj <- NormalizeMetacells(seurat_obj)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cee0960",
   "metadata": {},
   "outputs": [],
   "source": [
    "seurat_obj <- SetDatExpr(\n",
    "  seurat_obj,\n",
    "  group_name = unique(seurat_obj@meta.data$hash.ID), \n",
    "  group.by='hash.ID',\n",
    "  assay = 'RNA', # using RNA assay\n",
    "  slot = 'data' \n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a83ee9b",
   "metadata": {},
   "source": [
    "## Perform WGCNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3d41e9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test different soft powers:\n",
    "seurat_obj <- TestSoftPowers(\n",
    "  seurat_obj,\n",
    "  networkType = 'signed hybrid' \n",
    ")\n",
    "\n",
    "# plot the results:\n",
    "plot_list <- PlotSoftPowers(seurat_obj)\n",
    "\n",
    "# assemble with patchwork\n",
    "wrap_plots(plot_list, ncol=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "842bca9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "power_table <- GetPowerTable(seurat_obj)\n",
    "head(power_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0c3efe2",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "seurat_obj <- ConstructNetwork(\n",
    "  seurat_obj,\n",
    "  soft_power = 3,\n",
    "  consensus = FALSE,\n",
    "  overwrite_tom = TRUE,\n",
    "  wgcna_name = NULL,\n",
    "  blocks = NULL,\n",
    "  maxBlockSize = 30000 ,\n",
    "  randomSeed = 12345,\n",
    "  corType = \"pearson\",\n",
    "  consensusQuantile = 0.25,\n",
    "  networkType = \"signed hybrid\",\n",
    "  reassignThreshold = 0,\n",
    "  TOMType = \"signed\",\n",
    "  TOMDenom = \"min\",\n",
    "  scaleTOMs = TRUE,\n",
    "  scaleQuantile = 0.9,\n",
    "  sampleForScaling = TRUE,\n",
    "  sampleForScalingFactor = 1000,\n",
    "  useDiskCache = TRUE,\n",
    "  chunkSize = NULL,\n",
    "  deepSplit = 4,\n",
    "  pamStage = FALSE,\n",
    "  detectCutHeight = 0.995,\n",
    "  minModuleSize = 10,\n",
    "  mergeCutHeight = 0,\n",
    "  saveConsensusTOMs = TRUE\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0b814e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "PlotDendrogram(seurat_obj, main='LecMicro_human hdWGCNA Dendrogram')\n",
    "modules <- GetModules(seurat_obj) \n",
    "table((modules$module))\n",
    "\n",
    "length(unique(modules$module))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2adb78f1",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "seurat_obj <- ModuleEigengenes(\n",
    " seurat_obj,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77dfce4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "MEs <- GetMEs(seurat_obj, harmonized=FALSE)\n",
    "\n",
    "seurat_obj <- ModuleConnectivity(\n",
    "  seurat_obj\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6687e23c",
   "metadata": {},
   "outputs": [],
   "source": [
    "PlotKMEs <- function(\n",
    "  seurat_obj,\n",
    "  n_hubs=10,\n",
    "  text_size=2,\n",
    "  ncol = 5,\n",
    "  plot_widths = c(3,2),\n",
    "  wgcna_name = NULL\n",
    "){\n",
    "\n",
    "  if(is.null(wgcna_name)){wgcna_name <- seurat_obj@misc$active_wgcna}\n",
    "\n",
    "  modules <- GetModules(seurat_obj, wgcna_name) %>% subset(module != 'grey')\n",
    "  mods <- levels(modules$module); mods <- mods[mods != 'grey']\n",
    "  mod_colors <- modules %>% subset(module %in% mods) %>%\n",
    "    dplyr::select(c(module, color)) %>%\n",
    "    distinct\n",
    "\n",
    "\n",
    "  #get hub genes:\n",
    "  hub_df <- do.call(rbind, lapply(mods, function(cur_mod){\n",
    "    print(cur_mod)\n",
    "    cur <- subset(modules, module == cur_mod)\n",
    "    cur <- cur[,c('gene_name', 'module', paste0('kME_', cur_mod))]\n",
    "    names(cur)[3] <- 'kME'\n",
    "    cur <- dplyr::arrange(cur, kME)\n",
    "    top_genes <- cur %>% dplyr::top_n(n_hubs, wt=kME) %>% .$gene_name\n",
    "    cur$lab <- ifelse(cur$gene_name %in% top_genes, cur$gene_name, \"\")\n",
    "    cur\n",
    "  }))\n",
    "  head(hub_df)\n",
    "\n",
    "  plot_list <- lapply(mods, function(x){\n",
    "    print(x)\n",
    "    cur_color <- subset(mod_colors, module == x) %>% .$color\n",
    "    cur_df <- subset(hub_df, module == x)\n",
    "    top_genes <- cur_df %>% dplyr::top_n(n_hubs, wt=kME) %>% .$gene_name\n",
    "    p <- cur_df %>% ggplot(aes(x = reorder(gene_name, kME), y = kME)) +\n",
    "      geom_bar(stat='identity', width=1, color = cur_color, fill=cur_color) +\n",
    "      ggtitle(x) +\n",
    "      #xlab(paste0('kME_', x)) +\n",
    "      theme(\n",
    "        axis.ticks.x = element_blank(),\n",
    "        axis.text.x = element_blank(),\n",
    "        plot.title = element_text(hjust=0.5),\n",
    "        axis.title.x = element_blank(),\n",
    "        axis.line.x = element_blank(), \n",
    "          title = element_text(size = 11.5)\n",
    "      )\n",
    "    p_anno <- ggplot() + annotate(\n",
    "      \"label\",\n",
    "      x = 0,\n",
    "      y = 0,\n",
    "      label = paste0(top_genes, collapse=\"\\n\"),\n",
    "      size=text_size,\n",
    "      fontface = 'italic',\n",
    "      label.size=0\n",
    "    ) + theme_void()\n",
    "    patch <- p + p_anno + plot_layout(widths=plot_widths)\n",
    "    patch\n",
    "  })\n",
    "\n",
    "  wrap_plots(plot_list, ncol=ncol)\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "311a6bf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- PlotKMEs(seurat_obj, ncol=4, text_size = 3,   n_hubs=7)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eda6927f",
   "metadata": {},
   "outputs": [],
   "source": [
    "modules <- GetModules(seurat_obj) \n",
    "table((modules$module))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "581fbd49",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(modules, 'modules_hdWGCNA_lecMicro.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f2e4ba2",
   "metadata": {},
   "outputs": [],
   "source": [
    "micro.combined4 <- readRDS('./scrna_lecMicro_cleaned_integrated_sctransform.RDS')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85c86bdc",
   "metadata": {},
   "source": [
    "## Plot modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a065b2c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "ModuleNetworkPlot(\n",
    "  seurat_obj,\n",
    "  outdir = 'ModuleNetworks',  n_inner = 8,\n",
    "  n_outer = 12,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f59a75a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_names = GetHubGenes(seurat_obj, n_hubs =10)$gene_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09317d2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "modules <- GetModules(seurat_obj) %>% \n",
    "  subset(module != 'grey') %>% \n",
    "  mutate(module = droplevels(module))\n",
    "\n",
    "mods <- levels(modules$module)\n",
    "TOM <- GetTOM(seurat_obj)\n",
    "\n",
    "# get module colors for plotting \n",
    "mod_colors <- dplyr::select(modules, c(module, color)) %>% distinct()\n",
    "mod_cp <- mod_colors$color; names(mod_cp) <- as.character(mod_colors$module)\n",
    "\n",
    "# get genes in this pathway \n",
    "cur_genes <- gene_names\n",
    "\n",
    "# subset the TOM \n",
    "cur_TOM <- TOM[cur_genes,cur_genes] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce96667a",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph <- cur_TOM %>% \n",
    "  igraph::graph_from_adjacency_matrix(mode='undirected', weighted=TRUE) %>% \n",
    "  tidygraph::as_tbl_graph(directed=FALSE) %>% \n",
    "  tidygraph::activate(nodes) \n",
    "\n",
    "p <- ggraph(graph) + \n",
    "  geom_edge_link(color='grey', alpha=0.2) + \n",
    "  geom_node_point(color='black') +\n",
    "  geom_node_label(aes(label=name), repel=TRUE, max.overlaps=Inf, fontface='italic') \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a90cf9ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# only keep the upper triangular part of the TOM:\n",
    "cur_TOM[upper.tri(cur_TOM)] <- NA\n",
    "\n",
    "# cast the network from wide to long format\n",
    "cur_network <- cur_TOM %>% \n",
    "  reshape2::melt() %>% \n",
    "  dplyr::rename(gene1 = Var1, gene2 = Var2, weight=value) %>%\n",
    "  subset(!is.na(weight))\n",
    "\n",
    "# get the module & color info for gene1\n",
    "temp1 <- dplyr::inner_join(\n",
    "  cur_network,\n",
    "  modules %>% \n",
    "    dplyr::select(c(gene_name, module, color)) %>% \n",
    "    dplyr::rename(gene1 = gene_name, module1=module, color1=color),\n",
    "  by = 'gene1') %>% dplyr::select(c(module1, color1))\n",
    "\n",
    "# get the module & color info for gene2\n",
    "temp2 <- dplyr::inner_join(\n",
    "  cur_network,\n",
    "  modules %>% \n",
    "    dplyr::select(c(gene_name, module, color)) %>% \n",
    "    dplyr::rename(gene2 = gene_name, module2=module, color2=color),\n",
    "  by = 'gene2') %>% dplyr::select(c(module2, color2))\n",
    "\n",
    "# add the module & color info \n",
    "cur_network <- cbind(cur_network, temp1, temp2)\n",
    "\n",
    "# set the edge color to the module's color if they are the two genes are in the same module \n",
    "cur_network$edge_color <- ifelse(\n",
    "  cur_network$module1 == cur_network$module2, \n",
    "  as.character(cur_network$module1),\n",
    "  'grey')\n",
    "\n",
    "# keep this network before subsetting\n",
    "cur_network_full <- cur_network "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1efe8842",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=30, repr.plot.height=5)\n",
    "\n",
    "# subset to only keep edges between genes in the same module\n",
    "cur_network <- cur_network_full %>% \n",
    "  subset(module1 == module2)\n",
    "\n",
    "# make the graph object with tidygraph\n",
    "graph <- cur_network %>% \n",
    "  igraph::graph_from_data_frame() %>%\n",
    "  tidygraph::as_tbl_graph(directed=FALSE) %>% \n",
    "  tidygraph::activate(nodes)\n",
    "\n",
    "# add the module name to the graph:\n",
    "V(graph)$module <- modules[V(graph)$name,'module']\n",
    "\n",
    "# get the top 25 hub genes for each module\n",
    "hub_genes <- GetHubGenes(seurat_obj, n_hubs=25) %>% .$gene_name\n",
    "V(graph)$hub <- ifelse(V(graph)$name %in% hub_genes, V(graph)$name, \"\")\n",
    "\n",
    "# make the plot with gggraph\n",
    "p <- ggraph(graph) + \n",
    "  geom_edge_link(aes(alpha=weight, color=edge_color)) + \n",
    "  geom_node_point(aes(color=module)) +\n",
    "  geom_node_label(aes(label=hub, color = module), repel=TRUE, max.overlaps=Inf, fontface='italic') +\n",
    "  scale_colour_manual(values=mod_cp) +\n",
    "  scale_edge_colour_manual(values=mod_cp)+ theme_nothing() +\n",
    "  NoLegend() \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fefb302",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=4, repr.plot.height=39)\n",
    "\n",
    "\n",
    "plots <- list()\n",
    "\n",
    "\n",
    "interest = c(\"green\", \"magenta\", \"turquoise\", \"brown\", \"greenyellow\", \"blue\", \"pink\", \"black\", \"salmon\", \"red\", \"yellow\")\n",
    "# Replace this with your actual process of creating each graph\n",
    "for (i in 1:length(interest)) {\n",
    "  # Assuming `cur_network` is defined and updated for each graph creation\n",
    "  graph <- subset(cur_network, module1 == interest[i]) %>% \n",
    "    igraph::graph_from_data_frame() %>%\n",
    "    tidygraph::as_tbl_graph(directed = FALSE) %>% \n",
    "    tidygraph::activate(nodes)\n",
    "  \n",
    "  # Add the module name to the graph\n",
    "  V(graph)$module <- modules[V(graph)$name, 'module']\n",
    "  \n",
    "  # Get the top 25 hub genes for each module\n",
    "  hub_genes <- GetHubGenes(seurat_obj, n_hubs = 25) %>% .$gene_name\n",
    "  V(graph)$hub <- ifelse(V(graph)$name %in% hub_genes, V(graph)$name, \"\")\n",
    "  \n",
    "  # Make the plot with ggraph\n",
    "  p <- ggraph(graph) + \n",
    "    geom_edge_link(aes(alpha = weight, color = edge_color)) + \n",
    "    geom_node_point(aes(color = module, size = 3)) +\n",
    "    geom_node_label(aes(label = hub), repel = TRUE, max.overlaps = Inf, fontface = 'italic', size = 6) +\n",
    "    scale_colour_manual(values = mod_cp) +\n",
    "    scale_edge_colour_manual(values = mod_cp) +\n",
    "    theme_void() + # use theme_void() instead of theme_nothing() to avoid errors\n",
    "    theme(legend.position = \"none\") # use theme(legend.position = \"none\") instead of NoLegend()\n",
    "  \n",
    "  # Store the plot in the list\n",
    "  plots[[i]] <- p\n",
    "}\n",
    "\n",
    "plots <- lapply(plots, function(p) p + theme(plot.margin = margin(5, 5, 5, 5))) \n",
    "grid.arrange(grobs = plots, nrow = 11, ncol = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9f0f5ff",
   "metadata": {},
   "source": [
    "## Perform enrichment analysis of modules\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25f75a23",
   "metadata": {},
   "outputs": [],
   "source": [
    "# store modules as list \n",
    "\n",
    "# Initialize an empty list to store modules\n",
    "module_list <- list()\n",
    "\n",
    "# Loop through unique module names and extract genes associated with each module\n",
    "for (module in (modules$module)) {\n",
    "  module_genes <- modules$gene_name[modules$module == module]\n",
    "  module_list[[module]] <- module_genes\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4149c47",
   "metadata": {},
   "outputs": [],
   "source": [
    "#GO BPs enrichment \n",
    "\n",
    "results_BP = list()\n",
    "\n",
    "for (i in (unique(res$module))) {\n",
    "\n",
    "    gene_list <- (subset(modules, module == i))$gene_name\n",
    "    \n",
    "    go_enrich <- enrichGO(gene          = gene_list,\n",
    "                          universe     = row.names(micro.combined4),\n",
    "                          OrgDb        = org.Hs.eg.db,\n",
    "                          keyType      = \"SYMBOL\",\n",
    "                          ont          = \"BP\",   # Biological Process\n",
    "                          minGSSize = 10,\n",
    "                          pAdjustMethod = \"BH\",\n",
    "                          qvalueCutoff = 0.2)\n",
    "    \n",
    "    # Store the results in a list\n",
    "    results_BP[[i]] <- go_enrich\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c35d83f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#GO MFs enrichment \n",
    "\n",
    "results_MF = list()\n",
    "\n",
    "for (i in (unique(res$module))) {\n",
    "    \n",
    "\n",
    "    gene_list <- (subset(res, module == i))$gene_name\n",
    "    \n",
    "    # Perform GO enrichment analysis\n",
    "    go_enrich <- enrichGO(gene          = gene_list,\n",
    "                          universe     = row.names(micro.combined4),\n",
    "                          OrgDb        = org.Hs.eg.db,\n",
    "                          keyType      = \"SYMBOL\",\n",
    "                          ont          = \"MF\",   # Biological Process\n",
    "                        minGSSize = 10,\n",
    "                          pAdjustMethod = \"BH\",\n",
    "                          qvalueCutoff = 0.2)\n",
    "    \n",
    "    # Store the results in a list\n",
    "    results_MF[[i]] <- go_enrich\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2afe200d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_full <- data.frame()\n",
    "\n",
    "for (i in 1:length(results)){\n",
    "    df1 = results_BP[[i]]@result\n",
    "    df2=results_MF[[i]]@result\n",
    "    \n",
    "    df1$module = names(results)[[i]]\n",
    "    df2$module = names(results)[[i]]\n",
    "\n",
    "    df1$Ontology = (\"BP\")\n",
    "    df2$Ontology = (\"MF\")\n",
    "\n",
    "    df1$Description2 = paste(df1$Description, \"(BP)\")\n",
    "    df2$Description2 = paste(df2$Description, \"(MF)\")\n",
    "\n",
    "    df = rbind(df1, df2)\n",
    "    df <- df[order(-df$p.adjust, decreasing = TRUE), ]\n",
    "\n",
    "    df = head(df, 20)\n",
    "    GO_full = rbind(df_full, df)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46c26416",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "df_GO_save =GO_full[c('module','Ontology','ID','Description','GeneRatio','BgRatio',\n",
    "             'pvalue','p.adjust','qvalue','geneID','Count')]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcb32c45",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (i in 1:length(module_list)){\n",
    "    name = (names(module_list[i]))\n",
    "    name = paste(name,\".tab\", sep = \"\") \n",
    "    print(name)\n",
    "    assign(name, data.frame(geneID = ((module_list[i]))[[1]], set = rep(names(((module_list[i]))))))\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c10cea8",
   "metadata": {},
   "source": [
    "# ATM enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c926bc7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "law = (read.csv('../../ATM_genes/Lawrence2024_hs_syngo', header = 0))$V1\n",
    "ham = (read.csv('../../ATM_genes/Hammond2019_hs_syngo', header = 0))$V1\n",
    "li = (read.csv('../../ATM_genes/Li2019_hs_syngo', header = 0))$V1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e48dca8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to calculate overlaps and p-values using hypergeometric test\n",
    "calculate_overlap <- function(gene_list1, gene_list2, total_genes) {\n",
    "  overlap_genes <- length(intersect(gene_list1, gene_list2))\n",
    "  num_genes_list1 <- length(gene_list1)\n",
    "  num_genes_list2 <- length(gene_list2)\n",
    "\n",
    "  # Perform the hypergeometric test\n",
    "  p_value <- phyper(q = overlap_genes - 1, \n",
    "                    m = num_genes_list1, \n",
    "                    n = total_genes - num_genes_list1, \n",
    "                    k = num_genes_list2, \n",
    "                    lower.tail = FALSE)\n",
    "  \n",
    "  return(list(overlap = overlap_genes, p_val = p_value))\n",
    "}\n",
    "\n",
    "# Set the universe of genes (total genes in the dataset)\n",
    "total_genes <- length(row.names(micro.combined4))\n",
    "\n",
    "# Function to process each gene set and calculate overlaps for multiple modules\n",
    "process_gene_sets <- function(gene_list, module_list, set_name) {\n",
    "  overlap_list <- list()\n",
    "  p_val_list <- list()\n",
    "\n",
    "  for (i in 1:length(module_list)) {\n",
    "    gene_list2 <- module_list[[i]]\n",
    "    \n",
    "    # Calculate overlap and p-value\n",
    "    result <- calculate_overlap(gene_list, gene_list2, total_genes)\n",
    "    overlap_list[[names(module_list)[i]]] <- result$overlap\n",
    "    p_val_list[[names(module_list)[i]]] <- result$p_val\n",
    "  }\n",
    "\n",
    "  # Combine results into a data frame\n",
    "  overlap_df <- as.data.frame(t(as.data.frame(overlap_list)))\n",
    "  p_val_df <- as.data.frame(t(as.data.frame(p_val_list)))\n",
    "  overlap_df$p_val <- p_val_df$V1\n",
    "  overlap_df$set <- set_name\n",
    "  \n",
    "  # Set proper column names\n",
    "  colnames(overlap_df) <- c(\"overlap\", \"p.val\", \"set\")\n",
    "  return(overlap_df)\n",
    "}\n",
    "\n",
    "# Calculate overlaps for Li et al. and Lawrence et al.\n",
    "li_df <- process_gene_sets(li, module_list, \"li\")\n",
    "law_df <- process_gene_sets(law, module_list, \"law\")\n",
    "\n",
    "# Combine the two data frames\n",
    "df <- rbind(li_df, law_df)\n",
    "\n",
    "# Adjust p-values using BH correction\n",
    "df$p_adj <- p.adjust(df$p.val, method = \"BH\")\n",
    "\n",
    "# Restructure the data frame for final output\n",
    "df_li <- subset(df, set == \"li\")[c('overlap', 'p_adj')]\n",
    "colnames(df_li) <- c(\"Li_overlap\", \"Li_p.adj\")\n",
    "\n",
    "df_li$Lawrence_overlap <- subset(df, set == \"law\")$overlap\n",
    "df_li$Lawrence_p.adj <- subset(df, set == \"law\")$p_adj\n",
    "\n",
    "# Add the module names and reorder columns\n",
    "df_li$module <- row.names(df_li)\n",
    "df_li <- df_li[c(\"module\", \"Li_overlap\", \"Li_p.adj\", \"Lawrence_overlap\", \"Lawrence_p.adj\")]\n",
    "\n",
    "# Write the results to a file\n",
    "write.table(df_li, \"Lec_hdWGCNA_Modules_ATM_Overlap.tsv\", sep = '\\t', row.names = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d37cdcc7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4acd2cff",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Example sets\n",
    "set1 <- law\n",
    "set2 <- li\n",
    "set3 <- module_list[[\"pink\"]]\n",
    "\n",
    "\n",
    "# Calculate the areas for the 3-way Venn diagram\n",
    "venn.plot <- draw.triple.venn(\n",
    "  area1 = length(set1),\n",
    "  area2 = length(set2),\n",
    "  area3 = length(set3),\n",
    "  n12 = length(intersect(set1, set2)),\n",
    "  n13 = length(intersect(set1, set3)),\n",
    "  n23 = length(intersect(set2, set3)),\n",
    "  n123 = length(Reduce(intersect, list(set1, set2, set3))),\n",
    "  category = c(\"Lawrence\", \"Li\", \"Pink Module\"),\n",
    "  fill = c(\"red\", \"blue\", \"pink\"),\n",
    "  lty = \"dashed\",\n",
    "  cex = 2,\n",
    "  cat.cex = 2,\n",
    "  cat.col = c(\"red\", \"blue\", \"pink\")\n",
    ")\n",
    "\n",
    "# Display the Venn diagram\n",
    "grid.draw(venn.plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99ac51d5",
   "metadata": {},
   "source": [
    "## GSEA of modules in scRNA-seq (IgG1 vs. Lec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b902394d",
   "metadata": {},
   "outputs": [],
   "source": [
    "Idents(micro.combined4) <- micro.combined4@meta.data$Treatment\n",
    "DefaultAssay(micro.combined4) <- 'SCT'\n",
    "\n",
    "Lec_iGg1 = read.csv( '../110.ProcessFiles_B1B2/Lec_iGg1_DE_onlyFIRE_SCT_30APR2024.csv')\n",
    "row.names(Lec_iGg1) = Lec_iGg1$X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "124df0ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set plot dimensions\n",
    "options(repr.plot.width = 18, repr.plot.height = 10)\n",
    "\n",
    "# Function to create volcano plot\n",
    "volPlot <- function(markers, colors) {\n",
    "  \n",
    "  markers$X = rownames(markers)\n",
    "\n",
    "  ggplot(markers, aes(x = avg_log2FC, y = -log10(p_val_adj),  \n",
    "                      label = ifelse(p_val_adj < 0.05 & abs(avg_log2FC) > 0.05, X, \"\"),\n",
    "                      fill = factor(ifelse(\n",
    "                        avg_log2FC < 0 & p_val_adj < 0.05, \"Down-regulated\",\n",
    "                        ifelse(avg_log2FC > 0 & p_val_adj < 0.05, \"Up-regulated\", \"NS\")\n",
    "                      )))) +  \n",
    "\n",
    "    scale_fill_manual(name = \"Direction (padj < 0.05)\",\n",
    "                      values = colors) + \n",
    "    \n",
    "    geom_point(size = 5, shape = 21, color = \"grey2\", stroke = 0.2) +\n",
    "    \n",
    "    geom_hline(yintercept = 1.2, color = \"red\", linetype = \"dashed\") +\n",
    "    geom_vline(xintercept = 0, color = \"black\", linetype = \"dashed\") +\n",
    "    \n",
    "    # Styling for theme\n",
    "    theme_bw(base_size = 22) + \n",
    "    theme(\n",
    "      plot.title = element_text(hjust = 0.5, size = 24, face = \"bold\"),\n",
    "      axis.title = element_text(size = 18),\n",
    "      axis.text = element_text(size = 16),\n",
    "      legend.position = \"right\",\n",
    "      legend.title = element_text(size = 16),\n",
    "      legend.text = element_text(size = 14),\n",
    "      panel.grid = element_blank(),\n",
    "      panel.border = element_blank(),\n",
    "      strip.text = element_text(size = 16)\n",
    "    ) +\n",
    "    \n",
    "    # Add labels to significant genes\n",
    "    geom_label_repel(\n",
    "      size = 6, \n",
    "      max.overlaps = 30, \n",
    "      fill = NA, \n",
    "      force = 0.4, \n",
    "      nudge_y = -0.1, \n",
    "      box.padding = 0.5,\n",
    "      label.size = 3,\n",
    "      label.padding = 0.3\n",
    "    ) +\n",
    "    \n",
    "    # Set labels, title, legend\n",
    "    xlab(\"\\nAverage log2 Fold Change\") + \n",
    "    ylab(\"-log10(padj) \\n\") +\n",
    "    guides(\n",
    "      fill = guide_legend(order = 1),\n",
    "      col = guide_legend(order = 2)\n",
    "    )\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f1d9602",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=18, repr.plot.height=10)\n",
    "\n",
    "volPlot(Lec_iGg1, c(\"black\", \"grey\",\"red\"))  + ggtitle(\"Lec vs. iGg1 (magenta)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c31f897",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare gene set data\n",
    "els <- paste(names(module_list), \".tab\", sep = \"\")\n",
    "sets <- do.call(rbind, mget(els))\n",
    "colnames(sets) <- c(\"geneID\", \"set\")\n",
    "\n",
    "# Reorganize the sets data frame\n",
    "sets2 <- data.frame(set = sets$set, geneID = sets$geneID)\n",
    "\n",
    "# Prepare the gene list from your differential expression data\n",
    "df <- Lec_iGg1\n",
    "df$X <- rownames(df)\n",
    "original_gene_list <- df$avg_log2FC\n",
    "names(original_gene_list) <- df$X\n",
    "\n",
    "# Remove NA values and sort the gene list in decreasing order\n",
    "gene_list <- na.omit(original_gene_list)\n",
    "gene_list <- sort(gene_list, decreasing = TRUE)\n",
    "\n",
    "# Perform GSEA\n",
    "gsea_modules_sc <- GSEA(\n",
    "  gene_list, \n",
    "  TERM2GENE = sets2, \n",
    "  eps = 0, \n",
    "  minGSSize = 3, \n",
    "  maxGSSize = 3000,\n",
    "  pAdjustMethod = \"BH\",  \n",
    "  pvalueCutoff = 0.05\n",
    ")\n",
    "\n",
    "# Sort GSEA results by normalized enrichment score (NES)\n",
    "gsea_modules_sc@result <- gsea_modules_sc@result[order(gsea_modules_sc@result$NES), ]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0edeb395",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "options(repr.plot.width=11, repr.plot.height=10)\n",
    "\n",
    "# Assuming res_gsea$Description contains color names or hex codes\n",
    "res_gsea = gsea_modules_sc@result\n",
    "\n",
    "# Create the column reflecting NES /sig values\n",
    "res_gsea$Treatment = ifelse(res_gsea$NES > 0 & res_gsea$p.adjust < 0.05, \"Lec\", \n",
    "                   ifelse(res_gsea$NES < 0 & res_gsea$p.adjust < 0.05, \"iGg1\", \"C\"))\n",
    "\n",
    "\n",
    "res_gsea$OutlineColor <- ifelse(res_gsea$NES > 0, \"Lec\", \"IgG1\")\n",
    "res_gsea$OutlineColor <- factor(res_gsea$OutlineColor, levels = c(\"Lec\", \"IgG1\"))\n",
    "\n",
    "res_gsea$Significance <- ifelse(res_gsea$qvalue < 0.001, \"***\",\n",
    "                     ifelse(res_gsea$qvalue < 0.01, \"**\",\n",
    "                     ifelse(res_gsea$qvalue < 0.05, \"*\", \"\")))\n",
    "\n",
    "\n",
    "ggplot(res_gsea, aes(x = reorder(ID, -NES), y = NES, fill = Description, color = OutlineColor)) +\n",
    "      geom_bar(stat = \"identity\", width = 0.8, size = 0.8, alpha = 0.8) + \n",
    "      geom_text(aes(label = Significance, y = NES - 0.3 * sign(NES)), \n",
    "                size = 6, fontface = \"bold\") +\n",
    "    geom_hline(yintercept = 0,size = 1)+\n",
    "      scale_fill_identity() +  # Use colors from Description column\n",
    "      scale_color_manual(values = c(\"Lec\" = \"red2\", \"IgG1\" = \"black\")) +  \n",
    "      \n",
    "      scale_x_discrete(expand = expansion(add = .5)) +\n",
    "      scale_y_continuous(\n",
    "        breaks = seq(\n",
    "          floor(min(res_gsea$NES)), ceiling(max(res_gsea$NES)),\n",
    "          ceiling((ceiling(max(res_gsea$NES)) - floor(min(res_gsea$NES))) / 6)\n",
    "        )       ) +\n",
    "      coord_flip() + \n",
    "      theme_bw(base_size = 22) +   \n",
    "        theme(\n",
    "        legend.position = \"right\",\n",
    "        legend.key.size = unit(1.5, \"lines\"),\n",
    "        legend.text = element_text(size = 22),\n",
    "        axis.text = element_text(size = 22),\n",
    "        axis.title = element_text(size = 24), \n",
    "        plot.title = element_text(size = 24, face = 'bold', hjust = 0.6, \n",
    "                                 margin = margin(b = 20)), \n",
    "        plot.margin = margin(10,50,10, 50), \n",
    "        legend.title = element_text(size = 24)\n",
    "\n",
    "      ) +  \n",
    "      xlab(\"\")+\n",
    "    labs(title = \"GSEA Lec vs. IgG1: WGCNA modules\",  y = \"NES\", color = \"Direction\") + \n",
    "      labs(color = NULL, fill = NULL) +  \n",
    "    labs(fill = NULL) +  \n",
    "      guides(fill = \"none\", color = 'none')  \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r4_012024",
   "language": "R",
   "name": "r4_012024"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
